### PR TITLE
feature(github): cache github issues status in S3

### PR DESCRIPTION
**Pull request by fruch**
Ref: https://github.com/scylladb/qa-tasks/issues/1678

### Testing

- [ ]

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

File: .github/workflows/cache-issues.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+name: Cache issues status
+on:
+  schedule:
+    # Daily at 8:20 UTC
+    - cron: '20 8 * * *'
+  pull_request:
```
Comment (fruch) [inline .github/workflows/cache-issues.yaml:6]: remove pull_request and push, it was just for testing.

File: .github/workflows/cache-issues.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+name: Cache issues status
+on:
+  schedule:
+    # Daily at 8:20 UTC
+    - cron: '20 8 * * *'
```
Comment (fruch) [inline .github/workflows/cache-issues.yaml:5]: pick other time, also consider running it more then once a day

File: sdcm/utils/issues.py [hunk 3]
```diff
@@ -81,6 +105,18 @@ def get_issue_details(self, issue):
                                severity=Severity.WARNING,
                                trace=sys._getframe().f_back).publish()  # pylint: disable=protected-access
             return None
+        try:
+            if issue_details := self.cache.get_issue(owner=issue_parsed.user_id, repo_id=issue_parsed.repo_id, issue_id=issue_parsed.issue_id):
+                labels = [dict(name=label) for label in issue_details['labels'].strip().rstrip('|').split('|')]
```
Comment (fruch) [inline sdcm/utils/issues.py:110]: this can move to CachedGitHubIssues (that would be copied into dtest)

File: sdcm/utils/issues.py [hunk 4]
```diff
@@ -28,6 +30,26 @@ class Issue:
...
 
 
+class CachedGitHubIssues:
+    def __init__(self):
+        self.storage = KeyStore()
+
+    @lru_cache()
+    def get_repo_data(self, owner, repo):
+        scsv = self.storage.get_file_contents(f'issues/{owner}_{repo}.csv')
+        print('got repo')
```
Comment (soyacz) [inline sdcm/utils/issues.py:40]: debug leftover

File: sdcm/utils/issues.py (hunk 3 above)
Comment (soyacz) [reply sdcm/utils/issues.py:110]: +1

File: sdcm/utils/issues.py [hunk 5]
```diff
@@ -28,6 +30,26 @@ class Issue:
...
+        return {issue['id']: issue for issue in csv.DictReader(scsv.decode().splitlines(), fieldnames=("id", "state", "labels"))}
+
+    def get_issue(self, owner: str, repo_id: str, issue_id: str | int):
+        repo_issues_mapping = self.get_repo_data(owner, repo_id)
+        return repo_issues_mapping.get(str(issue_id))
+
+
+@lru_cache(maxsize=1)
+def get_issues_cache():
+    return CachedGitHubIssues()
```
Comment (soyacz) [inline sdcm/utils/issues.py:50]: can also make it singleton class

File: sdcm/utils/issues.py [hunk 6]
```diff
@@ -28,6 +30,26 @@ class Issue:
     issue_id: int | None
 
 
+class CachedGitHubIssues:
```
Comment (soyacz) [inline sdcm/utils/issues.py:33]: I'd add docstring how cache is populated - reference to gh action.

File: .github/workflows/cache-issues.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+          GH_TOKEN: ${{ secrets.ISSUE_ASSIGNMENT_TO_PROJECT_TOKEN }}
+      - name: Upload folder to bucket
+        uses: a-sync/s3-uploader@2.0.1
+        with:
+          args: --recursive
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.S3_AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.S3_AWS_SECRET_ACCESS_KEY }}
+          AWS_REGION: 'us-east-1'
+          S3_BUCKET: ${{ secrets.AWS_S3_BUCKET }}
```
Comment (soyacz) [inline .github/workflows/cache-issues.yaml:31]: make sure it's not readable by everyone

File: .github/workflows/cache-issues.yaml (hunk 2 above)
Comment (soyacz) [reply .github/workflows/cache-issues.yaml:5]: once per hour? IIRC limits are hour-based

File: .github/workflows/cache-issues.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+name: Cache issues status
```
Comment (soyacz) [inline .github/workflows/cache-issues.yaml:1]: is there some retry mechanism when it fails? 
Will we be informed when it fails and how?

**Comment by soyacz** [comment]
I wonder also about if this cache could be reused by Argus. In that case we might want to also cache also issue titles.

File: .github/workflows/cache-issues.yaml (hunk 2 above)
Comment (fruch) [reply .github/workflows/cache-issues.yaml:5]: I think one a day should be enough

we don't want thing to start running too soon, 
consider a test currently running with version that doesn't yet have a fix.

we don't need this accurate to the hour level, one day is enough.

File: sdcm/utils/issues.py (hunk 5 above)
Comment (fruch) [reply sdcm/utils/issues.py:50]: just used @cache on the class, it's exactly the same

File: .github/workflows/cache-issues.yaml (hunk 7 above)
Comment (fruch) [reply .github/workflows/cache-issues.yaml:31]: it's exactly where the rest of important info that SCT save, it's not public

File: .github/workflows/cache-issues.yaml (hunk 8 above)
Comment (fruch) [reply .github/workflows/cache-issues.yaml:1]: no retry, and it's visible on the actions page,

I think admin would get notification, but not sure.

File: sdcm/utils/issues.py (hunk 6 above)
Comment (fruch) [reply sdcm/utils/issues.py:33]: added one

File: sdcm/utils/issues.py (hunk 4 above)
Comment (fruch) [reply sdcm/utils/issues.py:40]: removed

File: sdcm/utils/issues.py (hunk 3 above)
Comment (fruch) [reply sdcm/utils/issues.py:110]: moved it

File: .github/workflows/cache-issues.yaml (hunk 8 above)
Comment (fruch) [reply .github/workflows/cache-issues.yaml:1]: we'll see, maybe as a follow-up we'll add a step to send mail notification,
but it's a bit more work to configure such a thing. (need to pass gmail cred in)

Comment (soyacz) [reply .github/workflows/cache-issues.yaml:1]: ok, let's see how it looks in practice and adapt as needed

**Review by soyacz — APPROVED** [review]
LGTM

File: .github/workflows/cache-issues.yaml (hunk 1 above)
Comment (fruch) [reply .github/workflows/cache-issues.yaml]: done

**Pull request merged**